File: simple_wallet.js
#!/usr/bin/env node
/**
 * Simple Reusable Wallet Manager
 * 
 * Creates and manages a permanent wallet address for ALGO transactions.
 */

const algosdk = require('algosdk');
const fs = require('fs');

// Algorand Testnet configuration
const ALGOD_TOKEN = '';
const ALGOD_SERVER = 'https://testnet-api.algonode.cloud';
const ALGOD_PORT = '';

// Create Algod client
const algodClient = new algosdk.Algodv2(ALGOD_TOKEN, ALGOD_SERVER, ALGOD_PORT);

const WALLET_FILE = 'permanent_wallet.json';

function createPermanentWallet() {
    console.log('🔑 Creating permanent wallet...');
    
    const account = algosdk.generateAccount();
    const address = account.addr;
    const mnemonic = algosdk.secretKeyToMnemonic(account.sk);
    
    const wallet = {
        address: address,
        mnemonic: mnemonic,
        privateKey: Array.from(account.sk),
        createdAt: new Date().toISOString(),
        type: 'permanent'
    };
    
    // Save wallet to file
    fs.writeFileSync(WALLET_FILE, JSON.stringify(wallet, null, 2));
    
    console.log('✅ Permanent wallet created!');
    console.log(`📍 Address: ${address}`);
    console.log(`🔐 Mnemonic: ${mnemonic}`);
    console.log(`💾 Saved to: ${WALLET_FILE}`);
    console.log('\n⚠️  IMPORTANT: Save this mnemonic safely!');
    console.log('   This is your permanent wallet address.');
    
    return wallet;
}

function loadPermanentWallet() {
    if (!fs.existsSync(WALLET_FILE)) {
        console.log('❌ No permanent wallet found. Creating new one...');
        return createPermanentWallet();
    }
    
    try {
        const walletData = fs.readFileSync(WALLET_FILE, 'utf8');
        const wallet = JSON.parse(walletData);
        
        // Handle different wallet file formats
        const address = wallet.address || wallet.addr;
        const createdAt = wallet.createdAt || wallet.note || 'Unknown';
        
        console.log('✅ Permanent wallet loaded!');
        console.log(`📍 Address: ${address}`);
        console.log(`📅 Created: ${createdAt}`);
        
        // Normalize wallet format
        return {
            address: address,
            addr: address, // Keep both for compatibility
            mnemonic: wallet.mnemonic,
            privateKey: wallet.privateKey,
            createdAt: createdAt,
            type: wallet.type || 'permanent'
        };
    } catch (error) {
        console.log('❌ Error loading wallet. Creating new one...');
        return createPermanentWallet();
    }
}

function getAccountFromWallet(wallet) {
    // If we have privateKey array, use it; otherwise derive from mnemonic
    let privateKey;
    if (wallet.privateKey && Array.isArray(wallet.privateKey)) {
        privateKey = new Uint8Array(wallet.privateKey);
    } else if (wallet.mnemonic) {
        const account = algosdk.mnemonicToSecretKey(wallet.mnemonic);
        privateKey = account.sk;
    } else {
        throw new Error('No private key or mnemonic found in wallet');
    }
    
    return {
        addr: wallet.address || wallet.addr,
        sk: privateKey
    };
}

async function checkBalance(address) {
    try {
        const accountInfo = await algodClient.accountInformation(address).do();
        const balance = Number(accountInfo.amount) / 1e6;
        return balance;
    } catch (error) {
        console.log('❌ Error checking balance:', error.message);
        return 0;
    }
}

async function fundWallet(wallet) {
    console.log('\n💰 Wallet Funding Instructions:');
    console.log(`📍 Your permanent wallet address: ${wallet.address}`);
    console.log('🔗 Go to: https://testnet.algoexplorer.io/dispenser');
    console.log('📝 Copy and paste your address above');
    console.log('💵 Request at least 10 ALGO (it\'s free!)');
    console.log('⏳ Press Enter when you have funded the wallet...');
    
    await new Promise(resolve => {
        process.stdin.once('data', () => resolve());
    });
    
    const balance = await checkBalance(wallet.address);
    console.log(`✅ Wallet balance: ${balance} ALGO`);
    
    if (balance < 1) {
        console.log('⚠️  Low balance! Please fund your wallet.');
        return false;
    }
    
    return true;
}

async function sendALGO(fromWallet, toAddress, amount) {
    try {
        console.log(`\n💸 Sending ${amount} ALGO to ${toAddress}...`);
        
        const account = getAccountFromWallet(fromWallet);
        const sp = await algodClient.getTransactionParams().do();
        
        // Convert to v2 format
        const suggestedParams = {
            fee: Number(sp.fee),
            firstRound: Number(sp.firstRound),
            lastRound: Number(sp.lastRound),
            genesisID: sp.genesisID,
            genesisHash: sp.genesisHash
        };
        
        const txn = algosdk.makePaymentTxnWithSuggestedParamsFromObject({
            from: account.addr,
            to: toAddress,
            amount: amount * 1e6, // Convert to microALGO
            suggestedParams: suggestedParams
        });
        
        const signedTxn = txn.signTxn(account.sk);
        const txId = await algodClient.sendRawTransaction(signedTxn).do();
        
        console.log(`📤 Transaction sent: ${txId}`);
        
        const result = await algosdk.waitForConfirmation(algodClient, txId, 4);
        console.log('✅ Transaction confirmed!');
        
        return txId;
    } catch (error) {
        console.error('❌ Error sending ALGO:', error.message);
        throw error;
    }
}

async function main() {
    try {
        console.log('🌟 Simple Reusable Wallet Manager\n');
        
        // Load or create permanent wallet
        const wallet = loadPermanentWallet();
        
        // Check current balance
        const balance = await checkBalance(wallet.address);
        console.log(`💰 Current balance: ${balance} ALGO`);
        
        if (balance < 1) {
            console.log('\n💳 Wallet needs funding...');
            const funded = await fundWallet(wallet);
            if (!funded) {
                console.log('❌ Please fund your wallet first.');
                return;
            }
        }
        
        console.log('\n🎯 Your Permanent Wallet:');
        console.log(`📍 Address: ${wallet.address}`);
        console.log(`💰 Balance: ${await checkBalance(wallet.address)} ALGO`);
        
        console.log('\n📋 Available Commands:');
        console.log('1. Check balance: node simple_wallet.js balance');
        console.log('2. Send ALGO: node simple_wallet.js send <address> <amount>');
        console.log('3. Fund wallet: node simple_wallet.js fund');
        
        // Handle command line arguments
        const args = process.argv.slice(2);
        
        if (args[0] === 'balance') {
            const balance = await checkBalance(wallet.address);
            console.log(`💰 Balance: ${balance} ALGO`);
        } else if (args[0] === 'send' && args[1] && args[2]) {
            const toAddress = args[1];
            const amount = parseFloat(args[2]);
            await sendALGO(wallet, toAddress, amount);
        } else if (args[0] === 'fund') {
            await fundWallet(wallet);
        } else {
            console.log('\n💡 Usage:');
            console.log('  node simple_wallet.js balance');
            console.log('  node simple_wallet.js send <address> <amount>');
            console.log('  node simple_wallet.js fund');
        }
        
    } catch (error) {
        console.error('❌ Error:', error.message);
    }
}

main();

File: simple_wallet.py
#!/usr/bin/env python3
"""
Simple Reusable Wallet Manager

Creates and manages a permanent wallet address for ALGO transactions.
"""

import json
import os
import sys
from datetime import datetime
from algosdk.v2client import algod
from algosdk import mnemonic, account, transaction
from algosdk.transaction import PaymentTxn

# Algorand Testnet configuration
ALGOD_TOKEN = ""
ALGOD_SERVER = "https://testnet-api.algonode.cloud"

# Create Algod client
algod_client = algod.AlgodClient(ALGOD_TOKEN, ALGOD_SERVER)

WALLET_FILE = "permanent_wallet.json"


def create_permanent_wallet():
    """Create a new permanent wallet"""
    print("🔑 Creating permanent wallet...")
    
    # Generate new account
    private_key, address = account.generate_account()
    wallet_mnemonic = mnemonic.from_private_key(private_key)
    
    wallet = {
        "address": address,
        "mnemonic": wallet_mnemonic,
        "privateKey": list(private_key),
        "createdAt": datetime.now().isoformat(),
        "type": "permanent"
    }
    
    # Save wallet to file
    with open(WALLET_FILE, 'w') as f:
        json.dump(wallet, f, indent=2)
    
    print("✅ Permanent wallet created!")
    print(f"📍 Address: {address}")
    print(f"🔐 Mnemonic: {wallet_mnemonic}")
    print(f"💾 Saved to: {WALLET_FILE}")
    print("\n⚠️  IMPORTANT: Save this mnemonic safely!")
    print("   This is your permanent wallet address.")
    
    return wallet


def load_permanent_wallet():
    """Load permanent wallet from file or create new one"""
    if not os.path.exists(WALLET_FILE):
        print("❌ No permanent wallet found. Creating new one...")
        return create_permanent_wallet()
    
    try:
        with open(WALLET_FILE, 'r') as f:
            wallet = json.load(f)
        
        # Handle different wallet file formats
        address = wallet.get('address') or wallet.get('addr')
        created_at = wallet.get('createdAt') or wallet.get('note', 'Unknown')
        
        print("✅ Permanent wallet loaded!")
        print(f"📍 Address: {address}")
        print(f"📅 Created: {created_at}")
        
        # Normalize wallet format
        return {
            "address": address,
            "addr": address,  # Keep both for compatibility
            "mnemonic": wallet['mnemonic'],
            "privateKey": wallet.get('privateKey'),
            "createdAt": created_at,
            "type": wallet.get('type', 'permanent')
        }
    
    except Exception as error:
        print("❌ Error loading wallet. Creating new one...")
        return create_permanent_wallet()


def get_account_from_wallet(wallet):
    """Get account object from wallet"""
    # If we have privateKey array, use it; otherwise derive from mnemonic
    if wallet.get('privateKey') and isinstance(wallet['privateKey'], list):
        private_key = bytes(wallet['privateKey'])
    elif wallet.get('mnemonic'):
        private_key = mnemonic.to_private_key(wallet['mnemonic'])
    else:
        raise ValueError('No private key or mnemonic found in wallet')
    
    return {
        "addr": wallet.get('address') or wallet.get('addr'),
        "sk": private_key
    }


def check_balance(address):
    """Check account balance"""
    try:
        account_info = algod_client.account_info(address)
        balance = account_info['amount'] / 1e6
        return balance
    except Exception as error:
        print(f"❌ Error checking balance: {error}")
        return 0


def fund_wallet(wallet):
    """Show funding instructions"""
    print("\n💰 Wallet Funding Instructions:")
    print(f"📍 Your permanent wallet address: {wallet['address']}")
    print("🔗 Go to: https://testnet.algoexplorer.io/dispenser")
    print("📝 Copy and paste your address above")
    print("💵 Request at least 10 ALGO (it's free!)")
    print("⏳ Press Enter when you have funded the wallet...")
    
    input()  # Wait for user input
    
    balance = check_balance(wallet['address'])
    print(f"✅ Wallet balance: {balance} ALGO")
    
    if balance < 1:
        print("⚠️  Low balance! Please fund your wallet.")
        return False
    
    return True


def send_algo(from_wallet, to_address, amount):
    """Send ALGO to another address"""
    try:
        print(f"\n💸 Sending {amount} ALGO to {to_address}...")
        
        account_obj = get_account_from_wallet(from_wallet)
        sp = algod_client.suggested_params()
        
        # Create payment transaction
        txn = PaymentTxn(
            sender=account_obj['addr'],
            sp=sp,
            receiver=to_address,
            amt=int(amount * 1e6)  # Convert to microALGO
        )
        
        # Sign transaction
        signed_txn = txn.sign(account_obj['sk'])
        
        # Send transaction
        tx_id = algod_client.send_transaction(signed_txn)
        print(f"📤 Transaction sent: {tx_id}")
        
        # Wait for confirmation
        result = transaction.wait_for_confirmation(algod_client, tx_id, 4)
        print("✅ Transaction confirmed!")
        
        return tx_id
    
    except Exception as error:
        print(f"❌ Error sending ALGO: {error}")
        raise error


def main():
    """Main function"""
    try:
        print("🌟 Simple Reusable Wallet Manager\n")
        
        # Load or create permanent wallet
        wallet = load_permanent_wallet()
        
        # Check current balance
        balance = check_balance(wallet['address'])
        print(f"💰 Current balance: {balance} ALGO")
        
        if balance < 1:
            print("\n💳 Wallet needs funding...")
            funded = fund_wallet(wallet)
            if not funded:
                print("❌ Please fund your wallet first.")
                return
        
        print("\n🎯 Your Permanent Wallet:")
        print(f"📍 Address: {wallet['address']}")
        print(f"💰 Balance: {check_balance(wallet['address'])} ALGO")
        
        print("\n📋 Available Commands:")
        print("1. Check balance: python simple_wallet.py balance")
        print("2. Send ALGO: python simple_wallet.py send <address> <amount>")
        print("3. Fund wallet: python simple_wallet.py fund")
        
        # Handle command line arguments
        args = sys.argv[1:]
        
        if len(args) == 0:
            print("\n💡 Usage:")
            print("  python simple_wallet.py balance")
            print("  python simple_wallet.py send <address> <amount>")
            print("  python simple_wallet.py fund")
        elif args[0] == "balance":
            balance = check_balance(wallet['address'])
            print(f"💰 Balance: {balance} ALGO")
        elif args[0] == "send" and len(args) >= 3:
            to_address = args[1]
            amount = float(args[2])
            send_algo(wallet, to_address, amount)
        elif args[0] == "fund":
            fund_wallet(wallet)
        else:
            print("\n💡 Usage:")
            print("  python simple_wallet.py balance")
            print("  python simple_wallet.py send <address> <amount>")
            print("  python simple_wallet.py fund")
    
    except Exception as error:
        print(f"❌ Error: {error}")


if __name__ == "__main__":
    main()
